### PR TITLE
tokens: make token controller regenerate invalid tokens

### PR DESCRIPTION
**Pull request by mikedanese**
If a token isn't valid against the token controller's signing key, the
token controller should regenerate the token.

Fixes: https://github.com/kubernetes/kubernetes/issues/62915

```release-note
The token controller now proactively rotates service account tokens when its signing key is changed.
```

### CODE REVIEW

File: pkg/controller/serviceaccount/tokens_controller_test.go [hunk 1]
```diff
@@ -568,124 +593,130 @@ func TestTokenCreation(t *testing.T) {
...
-		}
-
-		// This is the longest we'll wait for async tests
-		timeout := time.Now().Add(30 * time.Second)
-		waitedForAdditionalActions := false
-
-		for {
-			if controller.syncServiceAccountQueue.Len() > 0 {
-				controller.syncServiceAccount()
+		t.Run(k, func(t *testing.T) {
```
Comment (mikedanese) [inline pkg/controller/serviceaccount/tokens_controller_test.go]: This is the only change, everything else here is indentation.

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 2]
```diff
@@ -131,8 +136,9 @@ func NewTokensController(serviceAccounts informers.ServiceAccountInformer, secre
 
 // TokensController manages ServiceAccountToken secrets for ServiceAccount objects
 type TokensController struct {
-	client clientset.Interface
-	token  serviceaccount.TokenGenerator
+	client   clientset.Interface
+	tokgen   serviceaccount.TokenGenerator
```
Comment (awly) [inline pkg/controller/serviceaccount/tokens_controller.go]: tokGen, tokAuthN

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 3]
```diff
@@ -494,14 +500,15 @@ func (e *TokensController) hasReferencedToken(serviceAccount *v1.ServiceAccount)
...
 	caData := secret.Data[v1.ServiceAccountRootCAKey]
-	needsCA := len(e.rootCA) > 0 && bytes.Compare(caData, e.rootCA) != 0
+	needsCA = len(e.rootCA) > 0 && bytes.Compare(caData, e.rootCA) != 0
 
-	needsNamespace := len(secret.Data[v1.ServiceAccountNamespaceKey]) == 0
+	needsNamespace = len(secret.Data[v1.ServiceAccountNamespaceKey]) == 0
 
-	tokenData := secret.Data[v1.ServiceAccountTokenKey]
-	needsToken := len(tokenData) == 0
+	if _, ok, _ := e.tokauthn.AuthenticateToken(string(secret.Data[v1.ServiceAccountTokenKey])); !ok {
```
Comment (awly) [inline pkg/controller/serviceaccount/tokens_controller.go]:
```
_, tokenOK, _ := e.tokauthn.AuthenticateToken(string(secret.Data[v1.ServiceAccountTokenKey]))
needsToken = !tokenOK
```

Comment (awly) [reply pkg/controller/serviceaccount/tokens_controller.go]: Also maybe set needsToken = true on error?

File: pkg/controller/serviceaccount/tokens_controller_test.go [hunk 4]
```diff
@@ -22,23 +22,31 @@ import (
...
 	core "k8s.io/client-go/testing"
 	api "k8s.io/kubernetes/pkg/apis/core"
 	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/davecgh/go-spew/spew"
+	"gopkg.in/square/go-jose.v2/jwt"
 )
 
+var fakeAuthenticator = authenticator.TokenFunc(func(data string) (user.Info, bool, error) {
+	if len(data) == 0 || data == "invalid" {
```
Comment (awly) [inline pkg/controller/serviceaccount/tokens_controller_test.go]: `return nil, len(data) > 0 && data != "invalid", nil`

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 3 above)
Comment (mikedanese) [reply pkg/controller/serviceaccount/tokens_controller.go]: Done.

File: pkg/controller/serviceaccount/tokens_controller.go (hunk 2 above)
Comment (mikedanese) [reply pkg/controller/serviceaccount/tokens_controller.go]: Done.

File: pkg/controller/serviceaccount/tokens_controller_test.go (hunk 4 above)
Comment (mikedanese) [reply pkg/controller/serviceaccount/tokens_controller_test.go]: Done.

**Comment by liggitt** [comment]
A couple questions:
* Is this going to make initial start and resync really expensive on large clusters?
* Is it valid to assume that the controller manager has all the public keys in use by the apiserver authenticator? (I don't see how we're providing the full set to the controller manager)

**Comment by mikedanese** [comment]
> Is this going to make initial start and resync really expensive on large clusters?

Good question. I can test this out.

> Is it valid to assume that the controller manager has all the public keys in use by the apiserver authenticator? (I don't see how we're providing the full set to the controller manager)

It's not valid to assume that but I don't think we want to check against all valid public keys. e.g. if the apiserver is configured with an old and a new public key, and the controller is configured with the new private key, we want to rotate all tokens that aren't signed with the new key even though they might still be valid.

**Comment by liggitt** [comment]
> It's not valid to assume that but I don't think we want to check against all valid public keys. e.g. if the apiserver is configured with an old and a new public key…

I've also seen HA setups where controller managers are each given their own private key, and the three public keys are given to each apiserver. Not common, but it does exist.

**Comment by mikedanese** [comment]
> I've also seen HA setups where controller managers are each given their own private key, and the three public keys are given to each apiserver. Not common, but it does exist.

Would you be happy with this approach if we add a flag to specify valid verification keys and default to using the signing key?

**Comment by liggitt** [comment]
> > I've also seen HA setups where controller managers are each given their own private key, and the three public keys are given to each apiserver. Not common, but it does exist.
> 
> Would you be happy with this approach if we add a flag to specify valid verification keys and default to using the signing key?

Defaulting like that would make the controllers thrash by default. I think an option to give the controller the public keys, with no default, lets a deployment decide if they want the controller manager doing this rotation.

Were you able to determine the impact on startup/resync with large (O(100,000)) numbers of tokens

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 5]
```diff
@@ -494,14 +500,14 @@ func (e *TokensController) hasReferencedToken(serviceAccount *v1.ServiceAccount)
...
 	caData := secret.Data[v1.ServiceAccountRootCAKey]
-	needsCA := len(e.rootCA) > 0 && bytes.Compare(caData, e.rootCA) != 0
+	needsCA = len(e.rootCA) > 0 && bytes.Compare(caData, e.rootCA) != 0
 
-	needsNamespace := len(secret.Data[v1.ServiceAccountNamespaceKey]) == 0
+	needsNamespace = len(secret.Data[v1.ServiceAccountNamespaceKey]) == 0
 
-	tokenData := secret.Data[v1.ServiceAccountTokenKey]
-	needsToken := len(tokenData) == 0
+	_, tokenOk, err := e.tokAuthN.AuthenticateToken(string(secret.Data[v1.ServiceAccountTokenKey]))
```
Comment (enj) [inline pkg/controller/serviceaccount/tokens_controller.go]: If performance is an issue, would it be enough to just check that the key ID matches (and possibly the issuer)?

**Pull request closed without merging**